src/MaterialEditor.Base/NormalMapManager.cs

```diff
@@ -41,6 +41,9 @@ public bool ConvertNormalMap(ref Texture tex, string propertyName)
             if (!NormalMapProperties.Contains(propertyName))
                 return false;
 
+            if (tex == null || !tex.GetType().IsSubclassOf(typeof(Texture)))
+                return false;   // zipmod is broken ???   I used IsSubclassOf() because [tex is Texture] couldn't determine it.
+

```

Comment:
Since this is a very rare issue it might make more sense to use a try catch and catch the exception being thrown.

Reply:
I couldn't use try-catch in managed because it was crashing in native code.
Crashes should be avoided, even if they are rare.
I thought it was weird too, but tex.GetType() returned UnityEngine.Transform.

I made the check a function and added a comment.

Example that cannot be supplemented by enclosing with try-catch:
```
        public bool ConvertNormalMap(ref Texture tex, string propertyName)
        {
            System.Console.WriteLine($"@@@ ConvertNormalMap {propertyName}");
            try
            {
                if (!NormalMapProperties.Contains(propertyName))
                    return false;

                //if (tex == null || !tex.GetType().IsSubclassOf(typeof(Texture)))
                //    return false;   // zipmod is broken ???   I used IsSubclassOf() because [tex is Texture] couldn't determine it.
                
                ~~

                tex = normalTex;
                return true;
            }
            catch( System.Exception e )
            {
                System.Console.WriteLine(e);
                return false;
            }
        }
```

```
@@@ ConvertNormalMap NormalMap
Crash!!!
SymInit: Symbol-SearchPath: '.;C:\illusion\KoikatsuSunshine;C:\illusion\KoikatsuSunshine;C:\WINDOWS;C:\WINDOWS\system32;SRV*C:\websymbols*http://msdl.microsoft.com/download/symbols;', symOptions: 534, UserName: 'takah'
OS-Version: 10.0.0
~~
C:\WINDOWS\SYSTEM32\SspiCli.dll:SspiCli.dll (00007FFC65760000), size: 204800 (result: 0), SymType: '-deferred-', PDB: '', fileVersion: 10.0.19041.3636

========== OUTPUTTING STACK TRACE ==================

0x0000018D685DB7BB (Mono JIT Code) KK_Plugins.MaterialEditor.MaterialEditorPlugin:ConvertNormalMap (UnityEngine.Texture)
0x0000018D685DB53B (Mono JIT Code) MaterialEditorAPI.MaterialEditorPluginBase:ConvertNormalMap (UnityEngine.Texture&,string)
0x0000018D685DB193 (Mono JIT Code) KK_Plugins.MaterialEditor.MaterialEditorPlugin:ConvertNormalMap (UnityEngine.Material,string)
0x0000018D46250453 (Mono JIT Code) KK_Plugins.MaterialEditor.MaterialEditorPlugin:ConvertNormalMaps (UnityEngine.Material)
0x0000018D3D748BF3 (Mono JIT Code) KK_Plugins.MaterialEditor.MaterialEditorPlugin:AssetLoadedHook (XUnity.ResourceRedirector.AssetLoadedContext)
0x0000018D3D2B6BD6 (Mono JIT Code) XUnity.ResourceRedirector.ResourceRedirection:FireAssetLoadedEvent (XUnity.ResourceRedirector.AssetLoadedParameters,UnityEngine.AssetBundle,UnityEngine.Object[]&)
0x0000018D3D2B5FB3 (Mono JIT Code) XUnity.ResourceRedirector.ResourceRedirection:Hook_AssetLoaded_Postfix (XUnity.ResourceRedirector.AssetLoadingParameters,UnityEngine.AssetBundle,UnityEngine.Object[]&)
0x0000018D3D2B5E2B (Mono JIT Code) XUnity.ResourceRedirector.ResourceRedirection:Hook_AssetLoaded_Postfix (XUnity.ResourceRedirector.AssetLoadingParameters,UnityEngine.AssetBundle,UnityEngine.Object&)
0x0000018D3D2B5D1B (Mono JIT Code) XUnity.ResourceRedirector.Hooks.AssetBundle_LoadAsset_Internal_Hook:Postfix (UnityEngine.AssetBundle,UnityEngine.Object&,XUnity.ResourceRedirector.AssetLoadingContext&)
0x0000018D354EE4E3 (Mono JIT Code) (wrapper dynamic-method) MonoMod.Utils.DynamicMethodDefinition:NativeDetour_Wrapper<UnityEngine.AssetBundle::LoadAsset_Internal>?7 (UnityEngine.AssetBundle,string,System.Type)
0x0000018D3D2B4263 (Mono JIT Code) UnityEngine.AssetBundle:LoadAsset (string,System.Type)
0x0000018D3D2B59DB (Mono JIT Code) Sideloader.BundleManager:TryGetObjectFromName (string,string,System.Type,UnityEngine.Object&)
0x0000018D3D2B5333 (Mono JIT Code) Sideloader.Sideloader:RedirectHook (XUnity.ResourceRedirector.IAssetLoadingContext)
0x0000018D3D2B4A1B (Mono JIT Code) XUnity.ResourceRedirector.ResourceRedirection:Hook_AssetLoading_Prefix (XUnity.ResourceRedirector.AssetLoadingParameters,UnityEngine.AssetBundle,UnityEngine.Object[]&)
0x0000018D3D2B460B (Mono JIT Code) XUnity.ResourceRedirector.ResourceRedirection:Hook_AssetLoading_Prefix (XUnity.ResourceRedirector.AssetLoadingParameters,UnityEngine.AssetBundle,UnityEngine.Object&)
0x0000018D3D2B442B (Mono JIT Code) XUnity.ResourceRedirector.Hooks.AssetBundle_LoadAsset_Internal_Hook:Prefix (UnityEngine.AssetBundle,string&,System.Type&,UnityEngine.Object&,XUnity.ResourceRedirector.AssetLoadingContext&)
```
[output_log.txt](https://github.com/IllusionMods/KK_Plugins/files/14609323/output_log.txt)

Reply:
I think I understand the issue, Unity thinks the object is a Texture and casts it as such, but the underlying data is something else, which causes a crash in native. `is Texture` doesn't work because the object really is a Texture on the managed side, but it falls apart when calling anything on the native side. I'm not sure how `GetType` differs from `is` though, maybe it's overridden by Unity to pull from unmanaged data somehow, I don't know. It's fine as long as it works though.